File: gateway/internal/queue/amqp.go
// RabbitMQ 发布端：连接 + 幂等声明拓扑 + publisher confirm 发布。
// 与 Python 侧 worker/amqp.py、API 发布端用同一套拓扑名（前缀仅测试隔离用）。
package queue

import (
	"context"
	"errors"
	"fmt"
	"sync"

	amqp091 "github.com/rabbitmq/amqp091-go"

	"myink/gateway/internal/config"
)

// RabbitMQ 拓扑名（三端声明必须完全一致，参数不一致会 PRECONDITION_FAILED）。
const (
	ExchangeTasks = "myink.tasks"
	ExchangeDlx   = "myink.dlx"
	KeyTasks      = "tasks" // 主队列路由键
	KeyDelay      = "delay" // 延迟队列路由键
	KeyDlq        = "dlq"   // 死信路由键
)

// MainQueue / DelayQueue / DlqQueue 返回带前缀的队列名。QUEUE_PREFIX 只用于测试隔离，
// 生产为空串（exchange/queue 名与 Python 端一致）。
func MainQueue(prefix string) string  { return "queue:tasks" + prefix }
func DelayQueue(prefix string) string { return "queue:delay" + prefix }
func DlqQueue(prefix string) string   { return "queue:dlq" + prefix }

// ErrPublishAmbiguous 发布结果不确定（confirm 超时/连接中断）：消息可能已入队，不能补偿
// （worker 幂等兜底），但按失败返回给调用方。
var ErrPublishAmbiguous = fmt.Errorf("publish ambiguous")

// AMQP RabbitMQ 发布端。单 channel + mutex 串行发布（单用户项目量级远不需多 channel），
// channel 开 confirm 模式，每次发布等 broker 确认。TCP 被对端掐掉后 Publish 失败会重拨一次。
type AMQP struct {
	conn       *amqp091.Connection
	ch         *amqp091.Channel
	exchange   string
	routingKey string
	url        string
	prefix     string
	mu         sync.Mutex
}

// DialAMQP 连接 RabbitMQ 并幂等声明 5 项拓扑：
//
//	myink.tasks (direct)
//	  ├─ key "tasks" → queue:tasks  (x-max-priority=10, DLX→myink.dlx/dlq)
//	  └─ key "delay" → queue:delay  (DLX→myink.tasks/tasks；无固定 TTL，逐条 expiration 退避)
//	myink.dlx (direct) → key "dlq" → queue:dlq
//
// 失败返回 error（连接/声明/confirm 任一失败）。
func DialAMQP(cfg config.Config) (*AMQP, error) {
	a := &AMQP{
		exchange:   ExchangeTasks + cfg.QueuePrefix,
		routingKey: KeyTasks,
		url:        cfg.AmqpURL,
		prefix:     cfg.QueuePrefix,
	}
	if err := a.connect(); err != nil {
		return nil, err
	}
	return a, nil
}

func (a *AMQP) connect() error {
	if a.ch != nil {
		_ = a.ch.Close()
		a.ch = nil
	}
	if a.conn != nil {
		_ = a.conn.Close()
		a.conn = nil
	}
	conn, err := amqp091.Dial(a.url)
	if err != nil {
		return fmt.Errorf("amqp dial: %w", err)
	}
	ch, err := conn.Channel()
	if err != nil {
		_ = conn.Close()
		return fmt.Errorf("amqp channel: %w", err)
	}
	if err := ch.Confirm(false); err != nil {
		_ = ch.Close()
		_ = conn.Close()
		return fmt.Errorf("amqp confirm: %w", err)
	}
	a.conn = conn
	a.ch = ch
	if err := a.declare(a.prefix); err != nil {
		_ = a.Close()
		return err
	}
	return nil
}

// declare 幂等声明 5 项拓扑（重复声明安全；参数变更需清 RabbitMQ 数据卷）。
func (a *AMQP) declare(prefix string) error {
	exTasks := ExchangeTasks + prefix
	exDlx := ExchangeDlx + prefix
	mainQ, delayQ, dlqQ := MainQueue(prefix), DelayQueue(prefix), DlqQueue(prefix)

	if err := a.ch.ExchangeDeclare(exTasks, "direct", true, false, false, false, nil); err != nil {
		return fmt.Errorf("exchange %s: %w", exTasks, err)
	}
	// 主队列：优先级 + 死信 → myink.dlx/dlq
	if _, err := a.ch.QueueDeclare(mainQ, true, false, false, false, amqp091.Table{
		"x-max-priority":            int32(10),
		"x-dead-letter-exchange":    exDlx,
		"x-dead-letter-routing-key": KeyDlq,
	}); err != nil {
		return fmt.Errorf("queue %s: %w", mainQ, err)
	}
	if err := a.ch.QueueBind(mainQ, KeyTasks, exTasks, false, nil); err != nil {
		return fmt.Errorf("bind %s: %w", mainQ, err)
	}
	// 延迟队列：无固定 TTL，消息逐条带 expiration 死信回主队列实现退避（零插件）
	if _, err := a.ch.QueueDeclare(delayQ, true, false, false, false, amqp091.Table{
		"x-dead-letter-exchange":    exTasks,
		"x-dead-letter-routing-key": KeyTasks,
	}); err != nil {
		return fmt.Errorf("queue %s: %w", delayQ, err)
	}
	if err := a.ch.QueueBind(delayQ, KeyDelay, exTasks, false, nil); err != nil {
		return fmt.Errorf("bind %s: %w", delayQ, err)
	}
	// 死信交换机 + 死信队列
	if err := a.ch.ExchangeDeclare(exDlx, "direct", true, false, false, false, nil); err != nil {
		return fmt.Errorf("exchange %s: %w", exDlx, err)
	}
	if _, err := a.ch.QueueDeclare(dlqQ, true, false, false, false, nil); err != nil {
		return fmt.Errorf("queue %s: %w", dlqQ, err)
	}
	if err := a.ch.QueueBind(dlqQ, KeyDlq, exDlx, false, nil); err != nil {
		return fmt.Errorf("bind %s: %w", dlqQ, err)
	}
	return nil
}

// PublishTask 发布任务到主队列（持久化 + 优先级）。返回语义：
//   - nil：broker 已确认
//   - ErrPublishAmbiguous：confirm 超时/连接中断，结果不确定（消息可能已入队）
//   - 其他：确定失败（发送前错误 / broker nack），可安全补偿
func (a *AMQP) PublishTask(ctx context.Context, body []byte, priority int) error {
	a.mu.Lock()
	defer a.mu.Unlock()
	err := a.publishLocked(ctx, body, priority)
	if err == nil || errors.Is(err, ErrPublishAmbiguous) {
		return err
	}
	if rerr := a.connect(); rerr != nil {
		return fmt.Errorf("amqp reconnect: %w (publish: %v)", rerr, err)
	}
	return a.publishLocked(ctx, body, priority)
}

func (a *AMQP) publishLocked(ctx context.Context, body []byte, priority int) error {
	if a.ch == nil {
		return fmt.Errorf("amqp publish: channel closed")
	}
	dcf, err := a.ch.PublishWithDeferredConfirmWithContext(ctx, a.exchange, a.routingKey, false, false,
		amqp091.Publishing{
			ContentType:  "application/json",
			DeliveryMode: amqp091.Persistent,
			Priority:     uint8(priority),
			Body:         body,
		})
	if err != nil {
		return fmt.Errorf("amqp publish: %w", err)
	}
	ok, err := dcf.WaitContext(ctx)
	if err != nil {
		return ErrPublishAmbiguous
	}
	if !ok {
		return fmt.Errorf("amqp publish: broker nack")
	}
	return nil
}

// Close 关闭连接（通道随连接一起释放）。
func (a *AMQP) Close() error {
	if a.ch != nil {
		_ = a.ch.Close()
	}
	if a.conn != nil {
		return a.conn.Close()
	}
	return nil
}

File: src/myink/worker/amqp.py
"""RabbitMQ 连接与拓扑（worker 消费 + API/worker 发布共用）。

拓扑与 Go 网关 `gateway/internal/queue/amqp.go` 完全一致（producer/consumer/API 三端
幂等声明；参数不一致会 PRECONDITION_FAILED）：

    myink.tasks (direct)
      ├─ key "tasks" → queue:tasks  (x-max-priority=10, DLX→myink.dlx/dlq)
      └─ key "delay" → queue:delay  (DLX→myink.tasks/tasks；无固定 TTL，逐条 expiration 退避)
    myink.dlx (direct) → key "dlq" → queue:dlq

协议不变量（最高风险）：消息体 = 任务 JSON 本身（Go 端发布即任务字典，worker 单次
json.loads，不再包 {"body":...}）。

QUEUE_PREFIX 只用于测试隔离：对 exchange 与 queue 名统一加前缀（生产空串）。
"""

from __future__ import annotations

import threading

import pika

from myink.config import settings

EXCHANGE_TASKS = "myink.tasks"
EXCHANGE_DLX = "myink.dlx"
KEY_TASKS = "tasks"
KEY_DELAY = "delay"
KEY_DLQ = "dlq"


def _px() -> str:
    return settings.queue_prefix


def exchange() -> str:
    return EXCHANGE_TASKS + _px()


def dlx_exchange() -> str:
    return EXCHANGE_DLX + _px()


def main_queue() -> str:
    return "queue:tasks" + _px()


def delay_queue() -> str:
    return "queue:delay" + _px()


def dlq_queue() -> str:
    return "queue:dlq" + _px()


def connect() -> pika.BlockingConnection:
    """worker 消费连接：heartbeat=1800（批次任务可超 30 分钟默认看门狗，防通道被服务端杀掉）。

    返回 BlockingConnection（worker 单线程消费，非 asyncio 场景专用）。
    heartbeat/blocked_connection_timeout 是连接参数（非 URL 参数），URLParameters 只收 url。
    """
    params = pika.URLParameters(settings.amqp_url)
    params.heartbeat = 1800
    params.blocked_connection_timeout = None
    return pika.BlockingConnection(params)


def declare_topology(ch: pika.channel.Channel) -> None:
    """幂等声明 5 项拓扑（重复声明安全；参数变更需清 RabbitMQ 数据卷）。"""
    ch.exchange_declare(exchange(), exchange_type="direct", durable=True)
    # 主队列：优先级 + 死信 → myink.dlx/dlq
    ch.queue_declare(
        main_queue(),
        durable=True,
        arguments={
            "x-max-priority": 10,
            "x-dead-letter-exchange": dlx_exchange(),
            "x-dead-letter-routing-key": KEY_DLQ,
        },
    )
    ch.queue_bind(main_queue(), exchange(), KEY_TASKS)
    # 延迟队列：无固定 TTL，消息逐条带 expiration 死信回主队列实现退避（零插件）
    ch.queue_declare(
        delay_queue(),
        durable=True,
        arguments={
            "x-dead-letter-exchange": exchange(),
            "x-dead-letter-routing-key": KEY_TASKS,
        },
    )
    ch.queue_bind(delay_queue(), exchange(), KEY_DELAY)
    # 死信交换机 + 死信队列
    ch.exchange_declare(dlx_exchange(), exchange_type="direct", durable=True)
    ch.queue_declare(dlq_queue(), durable=True)
    ch.queue_bind(dlq_queue(), dlx_exchange(), KEY_DLQ)


# 模块级发布连接（publisher confirm）：API 路由线程池 / worker 重投共用，锁串行。
_publish_lock = threading.Lock()
_publish_conn: pika.BlockingConnection | None = None


def publish(body: str, routing_key: str, priority: int = 0, expiration_ms: int | None = None) -> None:
    """发布到 myink.tasks 交换机（持久化 + 优先级 + publisher confirm）。

    routing_key: KEY_TASKS（主队列）/ KEY_DELAY（延迟退避）。expiration_ms 非空时设置
    逐条 TTL（毫秒）——延迟队列 DLX 回主队列实现退避。
    阻塞式 confirm：broker nack / 连接断开抛 pika 异常；连接失效自动重建。
    """
    global _publish_conn
    with _publish_lock:
        try:
            if _publish_conn is None or _publish_conn.is_closed:
                _publish_conn = connect()
            ch = _publish_conn.channel()
            ch.confirm_delivery()  # 阻塞式 publisher confirm
            properties = pika.BasicProperties(delivery_mode=2, priority=priority)
            if expiration_ms is not None:
                properties.expiration = str(int(expiration_ms))
            ch.basic_publish(exchange=exchange(), routing_key=routing_key, body=body, properties=properties)
        except pika.exceptions.AMQPError:
            # 连接/通道失效：重建连接重试一次；仍失败抛给调用方（API 侧 503）
            try:
                if _publish_conn is not None and not _publish_conn.is_closed:
                    _publish_conn.close()
            except Exception:
                pass
            _publish_conn = connect()
            ch = _publish_conn.channel()
            ch.confirm_delivery()
            properties = pika.BasicProperties(delivery_mode=2, priority=priority)
            if expiration_ms is not None:
                properties.expiration = str(int(expiration_ms))
            ch.basic_publish(exchange=exchange(), routing_key=routing_key, body=body, properties=properties)
